### ISSUE TITLE
Proving the fibonacci example using `features = ["concurrent"]` results in panic

### ISSUE THREAD

**Issue by Dominik1999** [0xPolygonMiden/miden-vm#602]
I tried to benchmark the proving of the Fibonacci example with the default input file. It works fine when I single threadedly prove it. However, when I enable multi-threading by having 

```
miden = { version = "0.3", features = ["concurrent"]}
```

as a dependency, I get 
```
thread '<unnamed>' panicked at 'index out of bounds: the len is 4096 but the index is 20480', /Users/ich/.cargo/registry/src/github.com-1ecc6299db9ec823/winter-prover-0.4.2/src/constraints/evaluation_table.rs:287:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
thread '<unnamed>' panicked at 'index out of bounds: the len is 4096 but the index is 8192', /Users/ich/.cargo/registry/src/github.com-1ecc6299db9ec823/winter-prover-0.4.2/src/constraints/evaluation_table.rs:287:13
thread '<unnamed>' panicked at 'index out of bounds: the len is 4096 but the index is 4096', /Users/ich/.cargo/registry/src/github.com-1ecc6299db9ec823/winter-prover-0.4.2/src/constraints/evaluation_table.rs:287:13
thread 'thread '<unnamed><unnamed>' panicked at '' panicked at 'index out of bounds: the len is 4096 but the index is 24576index out of bounds: the len is 4096 but the index is 16384', ', /Users/ich/.cargo/registry/src/github.com-1ecc6299db9ec823/winter-prover-0.4.2/src/constraints/evaluation_table.rs/Users/ich/.cargo/registry/src/github.com-1ecc6299db9ec823/winter-prover-0.4.2/src/constraints/evaluation_table.rs::287287:13
:13
thread '<unnamed>' panicked at 'index out of bounds: the len is 4096 but the index is 28672', /Users/ich/.cargo/registry/src/github.com-1ecc6299db9ec823/winter-prover-0.4.2/src/constraints/evaluation_table.rs:287:13
thread '<unnamed>' panicked at 'index out of bounds: the len is 4096 but the index is 12288', /Users/ich/.cargo/registry/src/github.com-1ecc6299db9ec823/winter-prover-0.4.2/src/constraints/evaluation_table.rs:287:13
```

An easy way to reproduce is using https://github.com/0xPolygonMiden/examples/tree/main/benchmarking-cli and changing the `Cargo.toml`.

**Comment by bobbinth** [comment]
Could be related to this https://github.com/novifinancial/winterfell/issues/96 - but not 100% sure.

**Comment by 0xKanekiKen** [comment]
@bobbinth- yes, they are related. I've figured out the reason behind this issue. When running the code concurrently, we distribute the constraint evaluation work into multiple threads. During the evaluation of each fragment, we are adding an [offset](https://github.com/facebook/winterfell/blob/main/prover/src/constraints/evaluator.rs#L156) to the step which we are also passing when [updating transition evaluation row](https://github.com/facebook/winterfell/blob/main/prover/src/constraints/evaluator.rs#L156). We are seeing this error because the step(with offset) is greater than the length of each column. If we pass the index of the fragment row instead of the step, we won't see the code panicking.

**Comment by bobbinth** [comment]
@0xKanekiKen - great work! Seems like the fix in Winterfell should fix this too. I'll keep this open until we migrate to the next version of Winterfell.

**Comment by vlopes11** [comment]
Related to: https://github.com/0xPolygonMiden/miden-vm/issues/437

**Comment by vlopes11** [comment]
Fixed by #708
